Automattic/wp-calypso
# JPC: Retry auth

This PR implements a mechanism for retrying the jetpack authorization if it fails. We are observing a higher than normal number of users with problems connecting their jetpack sites due to a race condition when their site generates the secrets used in the communication with wp.com. Since this is a problem that, even if we solve in the jetpack plugin, could be with us for a long time (because of the jetpack release & update cycle), we need to do something from calypso to try to make the situation better.

Each time we try to authorize jetpack, we are generating a new set of secrets. So if the first attempt fails, we are going to try again with the new set of secrets that the new authorization generates.

For the users, the only thing they could notice is that calypso reloads several times (once per retry). If after the third attempt, the authorization keeps failing, we stop trying and fall back to the old error handling.

How to test
========
1. This is not easy to test because this bug can't be reproduced with a systematic approach. The easiest way would be to force error commenting the `if` condition in your jetpack plugin: 
https://github.com/Automattic/jetpack/blob/master/class.jetpack-xmlrpc-server.php#L195  : 

```
		//if ( ! hash_equals( $verify_secret, $secret_1 ) ) {
			Jetpack_Options::delete_option( $action );
			return $this->error( new Jetpack_Error( 'verify_secrets_mismatch', 'Secret mismatch', 400 ) );
		//}
```
You may need to do this change AFTER clicking on the connect button in your wp-admin, but before the calypso auth start. So be quick :)

2. Once calypso starts the auth dance, you should see the spinner and the "authorizing..." message for a while. It should reload calypso 3 times, and then it should end showing you the "secrets mismatched" error.
3. If you let a minute pass and reload, it should try again 
4. if you uncomment the class.jetpack-xmlrpc-server.php `if` condition you commented in 1) while calypso is trying to authorize the connection, it should complete it successfully


This is funny to test, uh?

ping @tyxla  @oskosk @dereksmart
